### DISCUSSION TITLE
CSV data extraction formatting from finance.yahoo.com

### DISCUSSION THREAD

**Discussion by umariani** [vlang/v#13071] [General]
CSV data can be download from finance.yahoo.com
     https://finance.yahoo.com/quote/AMZN/history
How can the raw data return be formatted to its csv format ?

Example

Date | Open | High | Low | Close* | Adj Close** | Volume
-- | -- | -- | -- | -- | -- | --
Jan 06, 2022 | 3,269.01 | 3,296.00 | 3,238.92 | 3,265.08 | 3,265.08 | 2,353,943
Jan 05, 2022 | 3,337.66 | 3,342.53 | 3,287.14 | 3,287.14 | 3,287.14 | 3,215,100
Jan 04, 2022 | 3,408.76 | 3,428.00 | 3,326.99 | 3,350.44 | 3,350.44 | 3,536,300

**Comment by Wajinn** [comment]
Possibly your question is a bit confusing.  Yahoo Finance already has a download link to put the data on the page into the .CSV format. In addition, CSV is just a text file.  You can rename text files separated in different ways with the .csv extension, and usually Excel, LibreOffice Calc, or whatever will ask you to identify the separator. 

Are you actually using an API instead?  If so, then you might want to show that as the link.  You could then import json and import net.http to start playing around with it.

Maybe refer to:
https://github.com/vlang/v/blob/master/examples/web_crawler/web_crawler.v
https://github.com/vlang/v/blob/master/examples/news_fetcher.v

**Reply by umariani** [reply]
Got this to work

=====================
CODE

import net.http

//=======================================================================
// finance.yahoo.com  Format to get history Quotes
//
//$urlA   = "https://query1.finance.yahoo.com/v7/finance/download/"
//symbol  = stock ticker                  // AMZN, NVDA
//$urlB   = "?period1="  + $Period1       // 1420848000
//$urlC   = "&period2="  + $Period2       // 1641859080
//$urlD   = "&interval=" + $DayWeekMonth  // 1wk
//$urlE   = "&events="   + $Events        // history
//$urlF   = "&crumb="    + myCrumb        // HMrC164NxHq
//
// ### Send COOKIE & URL+Crumb to Yahoo to fetch 1st stock history data,
// ### Cookie and AMZN
//=======================================================================

// Cannot use CamelCase for label -- Weird, make it harder to read

urla   := "https://query1.finance.yahoo.com/v7/finance/download/"
symbol := "AMZB"
urlb   := "?period1="  + "1420848000"
urlc   := "&period2="  + "1641859080"
urld   := "&interval=" + "1wk"
urle   := "&events="   + "history"
urlf   := "&crumb="    + "HMrC164NxHq"

url    := urla + symbol + urlb + urlc + urld + urle + urlf

// Send to Yahoo --- Looks like this ---
// URL: https://query1.finance.yahoo.com/v7/finance/download/AMZN?period1=1420848000&period2=1641859080&interval=1wk&events=history&crumb=HMrC164NxHq

println( url )

resp := http.get( url ) or {
println('Failed to fetch data from the server')
return
}
println( resp.text )


//==========================================
// Command to run
// c:\VLang\TEST>v run GetYahoo.v
//
// echo data sent to Yahoo
// https://query1.finance.yahoo.com/v7/finance/download/AMZN?period1=1420848000&period2=1641859080&interval=1wk&events=history&crumb=HMrC164NxHq
//
// OUTPUT
//
// Date,      Open,      High,      Low,       Close,     Adj Close, Volume
// 2015-01-05,301.480011,302.869995,296.679993,296.929993,296.929993,2592400
// 2015-01-12,297.559998,301.500000,285.250000,290.739990,290.739990,20993900
// 2015-01-19,292.589996,316.929993,286.390015,312.390015,312.390015,22999200
// 2015-01-26,311.820007,359.500000,299.329987,354.529999,354.529999,41666500
// 2015-02-02,350.049988,378.799988,350.010010,374.279999,374.279999,31754400
// 2015-02-09,371.000000,383.000000,367.200012,381.829987,381.829987,14064500
// ...
// ...
// 2021-12-06,3393.000000,3549.989990,3338.689941,3444.239990,3444.239990,14360700
// 2021-12-13,3440.000000,3483.320068,3303.899902,3400.350098,3400.350098,17017900
// 2021-12-20,3337.000000,3441.000000,3312.000000,3421.370117,3421.370117,10257600
// 2021-12-27,3420.739990,3458.860107,3331.169922,3334.340088,3334.340088,11720500
// 2022-01-03,3351.000000,3428.000000,3238.739990,3251.080078,3251.080078,14854600
// 2022-01-10,null,null,null,null,null,null
// 2022-01-10,3211.709961,3218.179932,3126.090088,3210.840088,3210.840088,3524802
//
// c:\VLang\TEST>
//
//==============================================
